// _datav-cli@0.2.1@datav-cli/static/js/config.js
/************* tab 页 ***************/
var configEditor   = require('./config_editor');
var dataEditor     = require('./data_editor');
var resizeEditor   = require('./resize_editor');
var publishEditor  = require('./publish_editor');
var testEditor  = require('./test_editor');
/************* tab 页 ***************/

module.exports = function (datav) {
  return {
    head: '<span class="chart-cnname" data-edit="false" data-input="<input type=\'text\' class=\'cn_ipt\' value=\'\'/>">' + datav['cn_name'] + '</span>&nbsp;<span class="chart-version">v' + window.share.version + '</span>',
    tabs: [
      {
        name: "配置",
        id: "config",
        content: "",
        showname: false,
        handler: configEditor
      },
      {
        name: "数据",
        id: "data",
        content: "",
        showname: true,
        handler: dataEditor
      },
      {
        name: "缩放",
        id: "resize",
        content: "",
        showname: true,
        handler: resizeEditor
      },
      {
        name: "测试",
        id: "test",
        content: "",
        showname: true,
        handler: testEditor
      },
      {
        name: "发布",
        id: "publish",
        content: "",
        showname: true,
        handler: publishEditor
      }
    ]
  }
}
