// datav-cli/static/js/publish_editor.js
var $ = require('jquery');
var _ = require('lodash');
var DatavGui = require('@ali/datav-gui');

var share = window.share;
var datav = share.datav;
var version = window.share.version;

var TYPE = {
  "请选择": '',
  '平面映射': 'map_plate',
  '球形映射': 'map_earth',
  '大标题': 'text_maintitle',
  '分组说明': 'text_category',
  '柱状图': 'regular_bar',
  '条形图': 'regular_column',
  '饼图': 'regular_pie',
  '折线图': 'regular_line',
  '区域图': 'regular_area',
  '散点图': 'regular_scatterplot',
  '树图': 'regular_tree',
  '力导向': 'network_force',
  '单一数字': 'figure_num',
  '翻牌板': 'figure_board',
  '监控指标': 'figure_monitor',
  '进度条': 'figure_progress',
  '简单排序': 'order_table',
  '分组排序': 'order_sort',
  '辅助图形': 'decorate'
}

module.exports = function (container) {
  var versionArr = version.split('.');
  versionArr[2] = parseInt(versionArr[2]) + 1;
  var publish_data = {
    view : {
      width : share.chartWidth,
      height: share.chartHeight,
      minWidth : share.chartMinWidth,
      minHeight: share.chartMinHeight
    },
    type : datav.type && datav.type[0] || '',
    version: versionArr.join('.'),
    icon: {
      '100x65': datav.icon && datav.icon['100x65'] || '',
      '55x55': datav.icon && datav.icon['55x55'] || ''
    }
  }
  var publishGui = new DatavGui.GUI({
    autoPlace: false
  });

  var viewCon = publishGui.addFolder('视窗');
  viewCon.add(publish_data.view, 'width').name('宽度');
  viewCon.add(publish_data.view, 'height').name('高度');
  viewCon.add(publish_data.view, 'minWidth').name('最小宽度');
  viewCon.add(publish_data.view, 'minHeight').name('最小高度');
  viewCon.open();
  publishGui.add(publish_data, 'type', TYPE).name('组件类型');

  var iconCon = publishGui.addFolder('图标');
  for (var key in publish_data.icon) {
    iconCon.addImage(publish_data.icon, key).req('/icon_save/' + key).res(function (data) {
      data = JSON.parse(data);
      if (data.isError) {
        alert('error : ' + data.message);
        return '/__static__/img/transparent.png';
      }
      return 'https://datav.oss-cn-hangzhou.aliyuncs.com/uploads/images/' + data.data;
    })
  }
  iconCon.open();
  publishGui.add(publish_data, 'version').name('发布版本');

  container.html(publishGui.domElement);
  container.append('<div class="publish-content-line content-line"><botton id="publish-btn" class="content-btn">发布</botton></div>');
  container.append('<p style="color:red">&nbsp;&nbsp;请注意版本格式：x.y.z   <small><a href="http://semver.org/lang/zh-CN/" target="_blank" style="color:red">原因详见</a></small></p>');
  container.append('<small>&nbsp;&nbsp;&nbsp;&nbsp;x位：API 修改，不再向下兼容</small><br/>');
  container.append('<small>&nbsp;&nbsp;&nbsp;&nbsp;y位：当你做了向下兼容的功能性新增，</small><br/>');
  container.append('<small>&nbsp;&nbsp;&nbsp;&nbsp;z位：当你做了向下兼容的问题修正。</small>');

  var publish_btn = $('#publish-btn');
  publish_btn.on('click', function () {
    if (publish_data.type === '') {
      alert('请选择图表类型');
      return;
    }
    var self = this;
    $(this).html('<image src="/__static__/iconfont/load.svg"></image>');
    $.post('/publish', publish_data, function (data) {
      $(self).html('发&nbsp;布');
      data = JSON.parse(data);
      if (data.isError) {
        alert('发布失败:' + data.message);
      } else {
        alert('发布成功!');
        location.reload();
      }
    })
  })
}
